### DISCUSSION TITLE
How to infer type from fetch on pages/api endpoint

### DISCUSSION THREAD

**Discussion by ZehuaZhang** [vercel/next.js#48497] [Help]
### Summary

In `page/api/example.ts`
```ts
interface MyRequest extends NextApiRequest {
}

export type MyResponseData = {
  
};

export default async (
  req: MyRequest,
  res: NextApiResponse<MyResponseData>
) => {
    // ....
    try {
        res.json(myData);
      } catch (error) {
        res
          .status(400)
          .end(JSON.stringify(error));
      }
};
```
How do I infer types for both data and error in client pages `fetch`?

**Comment by icyJoseph** [comment]
Hi,

As far as the fetch API goes, the return type is `any`. Many argue it should be `unknown`, but the fact remains that this is a matter of trust and verification. 

In this case, you control the data, as `MyResponseData`, and you should be able to trust that your fetch calls, when successful, return that data type. However, for errors, it gets a bit more complex, as the HTTP request could error at any point, some which might be out of your control.

> Some people use tools like zod/superstruct, client (😬 📦  careful with shipping to much to the client ❗) or server side, to validate/parse data.

One thing before moving on. The Error type when passed through `JSON.stringify` as you have it, resolves to. `'{}'`:

```js
let error = new Error("hello")
JSON.stringify(error) // '{}'
```

> I am a bit short on time, so apologies for any inaccuracies, omissions or lack of clarity.

A try/catch doesn't only catch errors, but rather thrown exceptions... 

```js
try {
 throw 3
} catch (reason) {
    console.log(reason) // 3 
}
```

So while I am not saying that you should cast the fetch result to the type, necessarily, you can:

- assume that if the fetch from your API endpoint is `ok`, you got the data you wanted
   - which in turn assumes assuming you verified/trust the data server side
- do some type of validation/assertion
   - Do remember that TypeScript types exist only on the code and are checked only at build time
   - Often the build step figures out that some code needs runtime type assertions
   - Unless you put these runtime checks in place, the compiler fails
   - It is rather up to the situation whether or not to type cast, but often it is a bad idea


```ts
// resolving to null for example's sake
const fetchData = async (): Promise<MyResponseData | null> => {
    try {
        const response = await fetch("/api/route");
        if (!response.ok) {
            throw new Error("Response was not ok") // some people throw the response entirely
        }
        const data = await response.json()
        // at this point, if you trust your data
        // then you should be able to proceed
        // if you don't trust your data, try to verify it server side first
        // if for some impossible reason you can't do that server side (maybe you don't own the API endpoint)
        // then you could create a function that validates/asserts the shape of data
        // validateOrAssertData(data) 

        return data
    } catch (reason) {
        // reason here is of unknown type
        if (reason instanceof Error) {
            // inside this block we know the reason to be of type Error
            console.log(reason.message)
        }
        // or maybe something else arrived here? an instance of Response?
 
        return null
    }
}
```

- https://kentcdodds.com/blog/get-a-catch-block-error-message-with-typescript

**Reply by ZehuaZhang** [reply]
I am looking for `infer` or `export type` or  `export interface` from server api to use on client side.
